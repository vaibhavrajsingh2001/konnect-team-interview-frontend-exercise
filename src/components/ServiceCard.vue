<template>
  <li class="service-card">
    <div class="service-card-header">
      <span>{{ statusParams.text }}</span>

      <div
        v-if="versionCount"
        class="version-badge"
      >
        {{ versionCount }} {{ versionCount > 1 ? 'versions' : 'version' }}
      </div>
    </div>
    <p class="service-name">
      {{ name }}
    </p>
    <p class="service-description">
      {{ description }}
    </p>

    <div class="service-metrics-and-avatars">
      <ServiceMetrics
        v-if="metrics"
        v-bind="metrics"
      />

      <k-avatar-group
        v-if="developers?.length"
        :avatar-list="developers"
      />
    </div>
  </li>
</template>
<script lang="ts">
import { defineComponent, type PropType } from 'vue'

import ServiceMetrics from '@/components/ServiceMetrics.vue'
import KAvatarGroup from '@/components/KAvatarGroup.vue'
import { ServiceStatus, type Developer, type Metrics } from '@/types'

export default defineComponent({
  name: 'ServiceCard',
  components: {
    ServiceMetrics,
    KAvatarGroup,
  },
  props: {
    name: {
      type: String,
      required: true,
    },
    description: {
      type: String,
      required: true,
    },
    versionCount: {
      type: Number,
      required: true,
    },
    published: {
      type: Boolean,
      required: true,
    },
    configured: {
      type: Boolean,
      required: true,
    },
    metrics: {
      type: Object as PropType<Metrics>,
      default: null,
    },
    developers: {
      type: Array as PropType<Developer[]>,
      default: () => [],
    },
  },
  computed: {
    // Computing the service status based on the published and configured props
    serviceStatus(): ServiceStatus {
      /**
       * - If the service is not published, return ServiceStatus.Unpublished
       * - If the service is published and configured, return ServiceStatus.Published
       * - If the service is published but not configured, return ServiceStatus.InProgress
       */
      return !this.published
        ? ServiceStatus.Unpublished
        : this.configured
          ? ServiceStatus.Published
          : ServiceStatus.InProgress
    },
    statusParams() {
      const statusParamMap = {
        [ServiceStatus.Unpublished]: {
          text: 'Unpublished',
          color: '#ADB6C8',
        },
        [ServiceStatus.Published]: {
          text: 'Published to portal',
          color: '#14B59A',
        },
        [ServiceStatus.InProgress]: {
          text: 'In progress',
          color: '#FABE5F',
        },
      }
      return statusParamMap[this.serviceStatus]
    },
  },
})
</script>

<style lang="scss">
.service-card {
  background-color: #FFFFFF;
  border-radius: 0.2rem;
  // Shadows make the card look elevated, hence better looking :)
  box-shadow: 0 1px 3px 0 rgb(0 0 0 / 0.1), 0 1px 2px -1px rgb(0 0 0 / 0.1);
  cursor: pointer;
  height: 24rem;
  padding: 2rem;
  width: 42rem;

  .service-card-header {
    align-items: center;
    display: flex;
    font-size: 1.2rem;
    font-weight: 400;
    height: 3.2rem;
    justify-content: space-between;
    line-height: 1.6rem;
    margin-bottom: 1rem;

    .version-badge {
      background-color: #F2F6FE;
      border-radius: 10rem;
      color: #5888DB;
      font-weight: 600;
      padding: 0.8rem 1.6rem;
    }
  }

  .service-name {
    font-size: 2rem;
    font-weight: 600;
    line-height: 2.4rem;
    margin-bottom: 1rem;
  }

  .service-description {
    -webkit-box-orient: vertical;
    color: #707888;
    display: -webkit-box;
    font-size: 1.3rem;
    font-weight: 400;
    -webkit-line-clamp: 1;
    line-height: 2rem;

    overflow: hidden;
  }

  .service-metrics-and-avatars {
    align-items: end;
    display: flex;
    justify-content: space-between;
  }
}
</style>
