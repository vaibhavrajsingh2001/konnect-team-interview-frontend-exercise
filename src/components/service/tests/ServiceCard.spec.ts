import { shallowMount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import ServiceCard from '@/components/service/ServiceCard.vue'

describe('ServiceCard', () => {
  const metrics = {
    latency: 0.5, uptime: 0.9278, requests: 248393, errors: 0.0244,
  }
  const developers = [
    { id: 'e06778e0-d298-4bc7-a2fe-db9c4b371f39', name: 'Oscar Brown', email: 'Oscar.Brown87@yahoo.com', avatar: 'https://avatars.githubusercontent.com/u/55118358' },
    { id: '2564a49a-2e3e-4dab-88c9-b10f895c1c84', name: 'Oscar Morissette', email: 'Oscar_Morissette57@gmail.com', avatar: 'https://avatars.githubusercontent.com/u/7590420' },
    { id: '71648bd0-1459-4d11-a698-abe58f4fefb4', name: 'Shaun Cronin', email: 'Shaun.Cronin91@gmail.com', avatar: 'https://avatars.githubusercontent.com/u/36033772' },
  ]

  it('renders correctly with no metrics or developers', () => {
    const wrapper = shallowMount(ServiceCard, {
      props: {
        name: 'Test Service',
        description: 'This is a test service',
        versionCount: 2,
        published: true,
        configured: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders correctly with metrics and developers', () => {
    const wrapper = shallowMount(ServiceCard, {
      props: {
        name: 'Test Service',
        description: 'This is a test service',
        versionCount: 1,
        published: false,
        configured: true,
        metrics,
        developers,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the correct service data', () => {
    const wrapper = shallowMount(ServiceCard, {
      props: {
        name: 'Test Service',
        description: 'This is a test service',
        versionCount: 2,
        published: true,
        configured: true,
      },
    })

    expect(wrapper.find('.service-card-header span').text()).toBe('Published to portal')
    expect(wrapper.find('.version-badge').text()).toBe('2 versions')
    expect(wrapper.find('.service-name').text()).toBe('Test Service')
    expect(wrapper.find('.service-description').text()).toBe('This is a test service')
  })

  it('renders the ServiceMetrics component when metrics prop is provided', () => {
    const wrapper = shallowMount(ServiceCard, {
      props: {
        name: 'Test Service',
        description: 'This is a test service',
        versionCount: 2,
        published: true,
        configured: true,
        metrics,
      },
    })

    expect(wrapper.findComponent({ name: 'ServiceMetrics' }).exists()).toBe(true)
  })

  it('renders the KAvatarGroup component when developers prop is provided', () => {
    const wrapper = shallowMount(ServiceCard, {
      props: {
        name: 'Test Service',
        description: 'This is a test service',
        versionCount: 2,
        published: true,
        configured: true,
        developers,
      },
    })

    expect(wrapper.findComponent({ name: 'KAvatarGroup' }).exists()).toBe(true)
  })
})
