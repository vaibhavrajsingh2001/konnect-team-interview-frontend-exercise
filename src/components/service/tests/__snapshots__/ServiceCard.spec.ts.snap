// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ServiceCard > renders correctly with metrics and developers 1`] = `
"<li class="service-card">
  <div class="service-card-header"><span class="service-status"><img src="/icons/cross.svg"> Unpublished</span>
    <div class="version-badge">1 version</div>
  </div>
  <p class="service-name">Test Service</p>
  <p class="service-description">This is a test service</p>
  <div class="service-metrics-and-avatars">
    <service-metrics-stub latency="0.5" uptime="0.9278" requests="248393" errors="0.0244"></service-metrics-stub>
    <k-avatar-group-stub avatarlist="[object Object],[object Object],[object Object]"></k-avatar-group-stub>
  </div>
</li>"
`;

exports[`ServiceCard > renders correctly with no metrics or developers 1`] = `
"<li class="service-card">
  <div class="service-card-header"><span class="service-status"><img src="/icons/check.svg"> Published to portal</span>
    <div class="version-badge">2 versions</div>
  </div>
  <p class="service-name">Test Service</p>
  <p class="service-description">This is a test service</p>
  <div class="service-metrics-and-avatars">
    <!--v-if-->
    <!--v-if-->
  </div>
</li>"
`;
